Feat: coinjoin middleware

karliatto opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Adds process for coinjoin-middleware https://github.com/trezor/coinjoin-backend

* Adds coinjoin binaries and file required at least for `linus-x64`
* New coinjoin process to manage running the binary
* Start coinjoin process form coinjoin module
* Update script for coinjoin middleware binaries (It is not completed since there are not releases yet)

Follow up issue for tests: https://github.com/trezor/trezor-suite/issues/6659
<!--- Describe your changes in detail -->

## Related Issue

Resolve  https://github.com/trezor/trezor-suite/issues/6420

karliatto commented on packages/suite-data/files/bin/coinjoin/linux-x64/WalletWasabi.WabiSabiClientLibrary.xml:
```diff
@@ -0,0 +1,162 @@
+<?xml version="1.0"?>
+<doc>
+    <assembly>
+        <name>WalletWasabi.WabiSabiClientLibrary</name>
+    </assembly>
+    <members>
+        <member name="M:WalletWasabi.WabiSabiClientLibrary.Controllers.CryptographyController.SelectUtxoForRound(WalletWasabi.WabiSabiClientLibrary.Models.SelectUtxoForRoundRequest)">
+            <summary>
+            Given a set of unspent transaction outputs, choose a subset of the outputs that are best to register in a single CoinJoin round according to the given strategy.
+            </summary>
```
Including this file is the only why I can make the binary work.
1. It is still not clear to me if we should include this files? 
2. Should it be included for all the binaries or only for `linux-x64`?
3. Should the binary be fix in a way that this file is not needed?

I created this issue https://github.com/trezor/coinjoin-backend/issues/11 in order to find the answers.

karliatto commented on packages/suite-desktop/src/libs/processes/CoinjoinProcess.ts:
```diff
@@ -0,0 +1,40 @@
+    async status(): Promise<Status> {
+        // service
+        try {
+            const resp = await fetch(`http://127.0.0.1:37128/Cryptography/analyze-transaction`, {
```
I do not see any sort of `status` end-point in the swagger documentation, so probably using this `analyze-transaction` is fine. Please let me know if other end-point is better.

szymonlesisz replied:
i guess there is no better endpoint, others requires more params

karliatto commented on packages/suite/src/services/coinjoin/config.ts:
```diff
@@ -10,6 +11,6 @@ export const COINJOIN_NETWORKS: Record<string, any> = {
         baseBlockHash: '0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206',
         // client settings
-        middlewareUrl: `${REGTEST_URL}Cryptography/`,
+        middlewareUrl: `${MIDDLEWARE_URL}Cryptography/`,
```
@szymonlesisz I assume this should be the only place where the URL for coinjoin middleware should be updated.

szymonlesisz replied:
lets leave it here as it is and  change it to valid url in suite-desktop/src/modules/coinjoin once we generate free port (see my comment above)

karliatto replied:
Thanks for the idea, it is now implemented as suggested.

karliatto commented on packages/suite-desktop/src/modules/coinjoin.ts:
```diff
@@ -85,6 +94,8 @@ export const init: Module = ({ mainWindow }) => {
         const dispose = () => {
             unregisterBackendProxy();
             unregisterClientProxy();
+            logger.info('coinjoin', 'Stopping (app quit)');
+            coinjoin.stop();
         };
```
Seams like good idea to stop it when app refreshes and quit. Since coinjoin account also disappears when app refreshes. 
With bridge we do different. We keep it running if app refreshes because it is for sure going to be needed. But it is not the case of coinjoin middleware.

dahaca replied:
The fact the CJ stays when a wallet is remembered is a feature or a bug? The process should be stopped regardless, I just wanted to clarify.

karliatto replied:
So when wallet is remembered the CJ account stays? I did not know that, if that is the case I think the process should not be stopped when refreshing, otherwise complicates the thing too much.
@szymonlesisz Could you clarify?

szymonlesisz replied:
it should be:
- disabled by suite (by calling "disable" method) while forgetting the accounts
- disabled on refresh, it will be enabled again when needed (by calling "createInstance")

mroz22 commented:
Nice, I guess we are not in hurry with merging it, we maybe even dont want to push it into the upcoming release since this is basically only privacy enhancement and we are going to release coinjoin feature only for internal purposes. So I guess we have plenty of  time for fine-tuning this.

Carlos, could you please create a similar test to what we have for bridge and tor processes? I could imagine that it would test happy path when  binary starts, and unhappy path, when binary for any reason does not. We could discuss that later.

karliatto commented:
> could you please create a similar test to what we have for bridge and tor processes?

Yes, that is a good idea, I will do that.
> I guess we are not in hurry with merging it

No need to hurry, but I think the sooner it gets tested the better.

szymonlesisz commented on packages/suite-desktop/src/libs/processes/CoinjoinProcess.ts:
```diff
@@ -0,0 +1,40 @@
+
+export class CoinjoinProcess extends BaseProcess {
+    constructor() {
+        super('coinjoin', 'WalletWasabi.WabiSabiClientLibrary', {
```
can we pass custom port? using getFreePort like we do with TOR?

usage:
```
WCL_BIND_PORT="1234" WalletWasabi.WabiSabiClientLibrary
```

karliatto replied:
Done.
I added default port that is basically always going to fail because when this class is initialized in `packages/suite-desktop/src/modules/coinjoin.ts` inside the `export const init: Module = ({ mainWindow }) => {` and modules are not loaded asynchronous so I thought it would be better idea to use `await getFreePort` everytime `onCreateInstance` is called. Because it is called asynchronously and that way we make sure we always get a free available port.

szymonlesisz commented on packages/suite-desktop/src/modules/coinjoin.ts:
```diff
@@ -51,11 +55,16 @@ export const init: Module = ({ mainWindow }) => {
                         const response = await client.enable();
-                        // TODO: start coinjoin middleware binary
-                        return response;
+                        try {
```
i cannot comment on unchanged lines.

lets generate free port on line 50, override `settings.middlewareUrl` to localhost:freeport and use those settings in constructor:
```
settings.middlewareUrl = 'http://localhost:{port}/Cryptography'
new CoinjoinClient(settings);
```

karliatto replied:
Done

szymonlesisz reviewed (changes requested):
> It is still not clear to me if we should include this files?
by adding to git-lfs, same as we do with TOR or Bridge
 
> Should it be included for all the binaries or only for linux-x64?
for all

> Should the binary be fix in a way that this file is not needed?
yes, we are trying to build it better (without swagger and unnecessary files like this xml)
so i guess this PR should be merged after we solve this build

karliatto commented:
@szymonlesisz
* Now binaries are included in git lfs fa9bd2d9619634837e31c14c1c43919b9df93324
* New binaries do not need to include the `.xml` file.
* The port for middleware is now dynamic from getFreePort util function,
* I rebased the whole thing.

szymonlesisz commented on packages/suite-desktop/src/modules/coinjoin.ts:
```diff
@@ -51,11 +58,16 @@ export const init: Module = ({ mainWindow }) => {
-                        // TODO: start coinjoin middleware binary
-                        return response;
+                        try {
+                            await coinjoin.startOnPort(port);
```
cool, i would change 2 things:
- throw error, pass it down to caller (suite) to be shown in suite toast
- add `method === 'disable'` case
```
onRequest: async (method, params) => {
        logger.debug(SERVICE_NAME, `CoinjoinClient call ${method}`);
        if (method === 'enable') {
            logger.debug(SERVICE_NAME, `CoinjoinClient binary enable on port ${port}`);
            try {
                await coinjoin.startOnPort(port);
            } catch (err) {
                logger.error(SERVICE_NAME, `Start failed: ${err.message}`);
                throw err; // pass this error to suite toast
            }
        }
        if (method === 'disable') {
            logger.debug(SERVICE_NAME, `CoinjoinClient binary stop`);
            coinjoin.stop();
        }
        // needs type casting
        return (client[method] as any)(...params); // bind method to instance context
    },
```

karliatto replied:
@szymonlesisz Ok
I understand that the code snippet that you pasted is pseudo-code and you didn't want to omit the  `const response = await client.enable();` part.

szymonlesisz replied:
nono, feel free to use my code should be working,

assigning response before calling `coinjoin.startOnPort` is not really necessary, it this fails we dont want to call `client.enable`

`client.enable` should be called when we are sure that everything is good

szymonlesisz commented on packages/suite-desktop/src/modules/coinjoin.ts:
```diff
@@ -85,6 +108,8 @@ export const init: Module = ({ mainWindow }) => {
         const dispose = () => {
             unregisterBackendProxy();
             unregisterClientProxy();
+            logger.info('coinjoin', 'Stopping (app quit)');
```
logger.info(SERVICE_NAME, 'Stopping (app quit)');

karliatto replied:
[9ec055c](https://github.com/trezor/trezor-suite/pull/6528/commits/9ec055ccbce6ea3852bb0bca66646f0b0a525497)

szymonlesisz reviewed (commented):
changes in [suite-desktop package.json](https://github.com/trezor/trezor-suite/pull/6485/commits/1c149fbc6874ce7f54fb7820d65f1af690e17f62) is required (your 3rd commit) in order to bundle binary into app

karliatto commented:
> changes in [suite-desktop package.json](https://github.com/trezor/trezor-suite/pull/6485/commits/1c149fbc6874ce7f54fb7820d65f1af690e17f62) is required (your 3rd commit) in order to bundle binary into app

[540b520](https://github.com/trezor/trezor-suite/pull/6528/commits/540b520e9884308ed6df0cc1a3abf1e598a04971)

Besides I am still investigating an issue I found that when it fails to run the binary the coinjoin client instance is still created so if you try again (after one failing) to create a coinjoin account it does not try to run the binary because the client is already created.

mroz22 commented on packages/suite-desktop/src/libs/processes/BaseProcess.ts:
```diff
@@ -133,13 +133,31 @@ export abstract class BaseProcess {
+                newProcess.removeListener('exit', spawnErrorHandler);
+                newProcess.removeListener('error', spawnErrorHandler);
+                resolve();
+            }, 100);
```
isn't there a better way than this timeout? 
- maybe this https://nodejs.org/api/child_process.html#event-spawn
- or maybe revisit `stdio: ['ignore', 'ignore', 'ignore']` and find a way how to consume stdout without killing app memory? and then wait for the first message hitting stdout?

szymonlesisz replied:
1. i've tried process.onspawn, it's fired **before** process.exit so it doeasnt help us
2. maybe instead of timeout lets constantly check `this.status()` in some time range, similar to how we awaiting trezor-user-env before test begins

karliatto replied:
I updated it to use `this.status()` and it works you can check it https://github.com/trezor/trezor-suite/pull/6528/commits/f0522182bd69ad3a84098c3f43f7762fc803e2f3

It makes it slower but it seams the right way. WDYT?

I rebased the whole thing.

karliatto commented:
@szymonlesisz 
* Updated middleware binaries to latest one
* I updated the commit messages to be more precised 
* and included the commit from where the middleware binaries come from

I am still investigating on how to do the `BaseProcess` in a  better way than the timeout.

karliatto commented:
@trezor/qa has tested app builds for all the platforms with successful result

linux - https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/3207936913
mac - https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/3207936910
win - https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/3207936915

szymonlesisz reviewed (approved):
perfect

mroz22 reviewed (approved):
works on my machine

matejkriz reviewed (changes requested):
not working on M1 machines and also tor is not starting (without any error message)

karliatto commented:
> not working on M1 machines and also tor is not starting (without any error message)

I created issue in the coinjoin-backend repository https://github.com/trezor/coinjoin-backend/issues/14

@matejkriz The Tor issue should be fix with fixup [baddbb2](https://github.com/trezor/trezor-suite/pull/6528/commits/baddbb2143006ee4fd0430108a9503ffeaf3634f)

matejkriz reviewed (approved):
On M1 mac you need to call
```bash
codesign -s - packages/suite-data/files/bin/coinjoin/osx-arm64/WalletWasabi.WabiSabiClientLibrary
```
to make it working in unsigned builds during development.

We could live with that temporary, but we need to have better solution later (signing the binary during build)
